incorrect peer type resolution taking the address of `catch |err| switch (err)`

### Zig Version

0.12

### Steps to Reproduce and Observed Behavior

```zig
test {
    @compileLog(@TypeOf(&(try @as(anyerror!usize, 0))));
    @compileLog(@TypeOf(&(@as(anyerror!usize, 0) catch {})));
    @compileLog(@TypeOf(&(@as(anyerror!usize, 0) catch |err| return err)));
    @compileLog(@TypeOf(&(@as(anyerror!usize, 0) catch |err| switch (err) {
        else => return err,
    })));
}
```
```
Compile Log Output:
@as(type, *const usize)
@as(type, *const usize)
@as(type, *const usize)
@as(type, usize)
```
this didnt happen in 0.11, im guessing its caused by #18173
interestingly if you take the address of the switch itself it becomes correct
```zig
@compileLog(@TypeOf((@as(anyerror!usize, 0) catch |err| &switch (err) {
    else => return err,
})));

```
`@as(type, *const usize)`

### Expected Behavior

all logs should be `*const usize`
```
Compile Log Output:
@as(type, *const usize)
@as(type, *const usize)
@as(type, *const usize)
@as(type, *const usize)
```

> this didnt happen in 0.11, im guessing its caused by #18173 interestingly if you take the address of the switch itself it becomes correct

Yes, that PR is the cause. The reason that taking the address does not exhibit the new (buggy) behaviour is that it does not make use of the new `switch_block_err_union` Zir instruction. One workaround (which is probably preferable to taking the address of the switch) is to store the result of the catch-switch expression in a decl before taking the address.

This issue is not dependent on having the type of  the switch be `noreturn`. Interestingly the (what should be) equivalent formulation using `if` (which also uses `switch_block_err_union`) does give the desired type:
```zig
test {
    const a: anyerror!usize = 0;
    const b = &(a catch |e| switch (e) {
        else => 3,
    });
    const c = &(if (a) |v| v else |e| switch (e) {
        else => 3,
    });
    @compileLog(@TypeOf(b));
    @compileLog(@TypeOf(c));
}
```
This results in 
```
bug.zig:9:5: error: found compile log statement
    @compileLog(@TypeOf(b));
    ^~~~~~~~~~~~~~~~~~~~~~~

Compile Log Output:
@as(type, usize)
@as(type, *const usize)
```
The Zir corresponding to these  two error union switches is:

```
%12 = switch_block_err_union(%8,
  non_err => {
    %13 = dbg_block_begin()
    %14 = err_union_payload_unsafe(%8) 
    %15 = dbg_block_end()
    %16 = break(%12, %14)
  },
  else => {
    %18 = dbg_block_begin()
    %19 = int(3)
    %20 = ref(%19) 
    %21 = dbg_block_end()
    %22 = break(%12, %20)
  }) 
%23 = dbg_var_val(%12, "b")
```
and
```
%26 = switch_block_err_union(%8,
  non_err => {
    %27 = dbg_block_begin()
    %28 = err_union_payload_unsafe(%8) 
    %31 = ref(%28) 
    %29 = dbg_block_begin()
    %30 = dbg_var_val(%28, "v")
    %32 = dbg_block_end()
    %33 = break(%26, %31)
},
  else => {
    %35 = dbg_block_begin()
    %36 = int(3)
    %37 = ref(%36) 
    %38 = dbg_block_end()
    %39 = break(%26, %37)
  }) 
%40 = dbg_var_val(%26, "c")
```
So the issue is that the `catch` version is not generating a `ref(%14)` in the `non_err` branch.

--- a/lib/std/zig/AstGen.zig
+++ b/lib/std/zig/AstGen.zig
@@ -7071,8 +7071,10 @@ fn switchExprErrUnion(
         .ctx = ri.ctx,
     };
 
-    const payload_is_ref = node_ty == .@"if" and
-        if_full.payload_token != null and token_tags[if_full.payload_token.?] == .asterisk;
+    const payload_is_ref = switch (node_ty) {
+        .@"if" => if_full.payload_token != null and token_tags[if_full.payload_token.?] == .asterisk,
+        .@"catch" => ri.rl == .ref or ri.rl == .ref_coerced_ty,
+    };
 
     // We need to call `rvalue` to write through to the pointer only if we had a
     // result pointer and aren't forwarding it.

--- a/test/behavior.zig
+++ b/test/behavior.zig
@@ -89,6 +89,7 @@ test {
     _ = @import("behavior/switch.zig");
     _ = @import("behavior/switch_prong_err_enum.zig");
     _ = @import("behavior/switch_prong_implicit_cast.zig");
+    _ = @import("behavior/switch_on_captured_error.zig");
     _ = @import("behavior/this.zig");
     _ = @import("behavior/threadlocal.zig");
     _ = @import("behavior/truncate.zig");

--- a/test/behavior/switch_on_captured_error.zig
+++ b/test/behavior/switch_on_captured_error.zig
@@ -3,6 +3,7 @@ const assert = std.debug.assert;
 const expect = std.testing.expect;
 const expectError = std.testing.expectError;
 const expectEqual = std.testing.expectEqual;
+const builtin = @import("builtin");
 
 test "switch on error union catch capture" {
     if (builtin.zig_backend == .stage2_spirv64) return error.SkipZigTest;

--- a/test/behavior/switch_on_captured_error.zig
+++ b/test/behavior/switch_on_captured_error.zig
@@ -18,6 +18,7 @@ test "switch on error union catch capture" {
             try testCapture();
             try testInline();
             try testEmptyErrSet();
+            try testAddressOf();
         }
 
         fn testScalar() !void {
@@ -254,6 +255,44 @@ test "switch on error union catch capture" {
                 try expectEqual(@as(u64, 0), b);
             }
         }
+
+        fn testAddressOf() !void {
+            {
+                const a: anyerror!usize = 0;
+                const ptr = &(a catch |e| switch (e) {
+                    else => 3,
+                });
+                comptime assert(@TypeOf(ptr) == *const usize);
+                try expectEqual(ptr, &(a catch unreachable));
+            }
+            {
+                const a: anyerror!usize = error.A;
+                const ptr = &(a catch |e| switch (e) {
+                    else => 3,
+                });
+                comptime assert(@TypeOf(ptr) == *const comptime_int);
+                try expectEqual(3, ptr.*);
+            }
+            {
+                var a: anyerror!usize = 0;
+                _ = &a;
+                const ptr = &(a catch |e| switch (e) {
+                    else => return,
+                });
+                comptime assert(@TypeOf(ptr) == *usize);
+                ptr.* += 1;
+                try expectEqual(@as(usize, 1), a catch unreachable);
+            }
+            {
+                var a: anyerror!usize = error.A;
+                _ = &a;
+                const ptr = &(a catch |e| switch (e) {
+                    else => return,
+                });
+                comptime assert(@TypeOf(ptr) == *usize);
+                unreachable;
+            }
+        }
     };
 
     try comptime S.doTheTest();
@@ -279,6 +318,7 @@ test "switch on error union if else capture" {
             try testInlinePtr();
             try testEmptyErrSet();
             try testEmptyErrSetPtr();
+            try testAddressOf();
         }
 
         fn testScalar() !void {
@@ -748,6 +788,45 @@ test "switch on error union if else capture" {
                     else => |e| return e,
                 };
                 try expectEqual(@as(u64, 0), b);
+            }
+        }
+
+        fn testAddressOf() !void {
+            if (builtin.zig_backend == .stage2_wasm) return error.SkipZigTest;
+            {
+                const a: anyerror!usize = 0;
+                const ptr = &(if (a) |*v| v.* else |e| switch (e) {
+                    else => 3,
+                });
+                comptime assert(@TypeOf(ptr) == *const usize);
+                try expectEqual(ptr, &(a catch unreachable));
+            }
+            {
+                const a: anyerror!usize = error.A;
+                const ptr = &(if (a) |*v| v.* else |e| switch (e) {
+                    else => 3,
+                });
+                comptime assert(@TypeOf(ptr) == *const comptime_int);
+                try expectEqual(3, ptr.*);
+            }
+            {
+                var a: anyerror!usize = 0;
+                _ = &a;
+                const ptr = &(if (a) |*v| v.* else |e| switch (e) {
+                    else => return,
+                });
+                comptime assert(@TypeOf(ptr) == *usize);
+                ptr.* += 1;
+                try expectEqual(@as(usize, 1), a catch unreachable);
+            }
+            {
+                var a: anyerror!usize = error.A;
+                _ = &a;
+                const ptr = &(if (a) |*v| v.* else |e| switch (e) {
+                    else => return,
+                });
+                comptime assert(@TypeOf(ptr) == *usize);
+                unreachable;
             }
         }
     };

--- a/test/behavior/switch_on_captured_error.zig
+++ b/test/behavior/switch_on_captured_error.zig
@@ -7,6 +7,7 @@ const builtin = @import("builtin");
 
 test "switch on error union catch capture" {
     if (builtin.zig_backend == .stage2_spirv64) return error.SkipZigTest;
+    if (builtin.zig_backend == .stage2_wasm) return error.SkipZigTest;
 
     const S = struct {
         const Error = error{ A, B, C };
@@ -261,6 +262,7 @@ test "switch on error union catch capture" {
 
 test "switch on error union if else capture" {
     if (builtin.zig_backend == .stage2_spirv64) return error.SkipZigTest;
+    if (builtin.zig_backend == .stage2_wasm) return error.SkipZigTest;
 
     const S = struct {
         const Error = error{ A, B, C };

